src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpAssemblyDocumentGenerator.cs

```diff
@@ -0,0 +1,370 @@
+using Microsoft.DotNet.ApiSymbolExtensions.Filtering;
+using Microsoft.DotNet.ApiSymbolExtensions.Logging;
+using Microsoft.DotNet.GenAPI.SyntaxRewriter;
+
+namespace Microsoft.DotNet.GenAPI;
+
+/// <summary>
+/// A class that generates the C# document and syntax trees of a specified collection of assemblies.
+/// </summary>
+public sealed class CSharpAssemblyDocumentGenerator
+{
+    private readonly ILog _log;
+    private readonly IAssemblySymbolLoader _loader;
+    private readonly ISymbolFilter _symbolFilter;
+    private readonly ISymbolFilter _attributeDataSymbolFilter;
+    private readonly string? _exceptionMessage;
+    private readonly bool _includeAssemblyAttributes;
+    private readonly AdhocWorkspace _adhocWorkspace;
+    private readonly SyntaxGenerator _syntaxGenerator;
+    private readonly IEnumerable<MetadataReference>? _metadataReferences;
+    private readonly bool _addPartialModifier;
+    private readonly bool _hideImplicitDefaultConstructors;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CSharpAssemblyDocumentGenerator"/> class.
+    /// </summary>
+    /// <param name="log">The logger to use.</param>
+    /// <param name="loader">The assembly symbol loader to use.</param>
+    /// <param name="symbolFilter">The symbol filter to use.</param>
+    /// <param name="attributeDataSymbolFilter">The attribute data symbol filter to use.</param>
+    /// <param name="exceptionMessage">The optional exception message to use.</param>
+    /// <param name="includeAssemblyAttributes">Whether to include assembly attributes or not.</param>
+    /// <param name="metadataReferences">The metadata references to use. The default value is <see langword="null"/>.</param>
+    /// <param name="addPartialModifier">Whether to add the partial modifier or not. The default value is <see langword="true"/>.</param>
+    /// <param name="hideImplicitDefaultConstructors">Whether to hide implicit default constructors or not. The default value is <see langword="true"/>.</param>
+    public CSharpAssemblyDocumentGenerator(ILog log,
+                                 IAssemblySymbolLoader loader,
+                                 ISymbolFilter symbolFilter,
+                                 ISymbolFilter attributeDataSymbolFilter,
+                                 string? exceptionMessage,
+                                 bool includeAssemblyAttributes,
+                                 IEnumerable<MetadataReference>? metadataReferences = null,
+                                 bool addPartialModifier = true,
+                                 bool hideImplicitDefaultConstructors = true)
+    {
+        _log = log;
+        _loader = loader;
+        _symbolFilter = symbolFilter;
+        _attributeDataSymbolFilter = attributeDataSymbolFilter;
+        _exceptionMessage = exceptionMessage;
+        _includeAssemblyAttributes = includeAssemblyAttributes;
+        _adhocWorkspace = new AdhocWorkspace();
```

Comment:
Is there a reason you can't pass this is as a parameter?

Reply:
So the code in this file was moved out of `CSharpFileBuilder.cs` (line 54 left side) and into this new class. I tried to leave the logic untouched.

I don't think the object needs to be passed from outside, just needs to be created during construction.

Reply:
I see Michael's point here. If you want to unit test this type you want to make sure that all dependencies are provided externally so that you can mock them. By having the construction of `AdhocWorkspace` inside this type, you can't mock it out and automatically execute business code when testing the type itself.

This is the general pattern that we also follow in APICompat to be able to only test only the unit of code.

Reply:
Good to keep in mind. But since this wasn't code I changed, and it was being used the same way before, I think we should consider it out of scope of this PR.